Requirements.md

```diff
@@ -0,0 +1,121 @@
+   
+1. Security is more important than architectural integrity.
+
+1. Architectural integrity is more important than performance.
+   'Architectural Integrity' is intended to denote simplicity/coherence of design. Generally speaking, architectural integrity can be measured in terms of the number of concepts and the complexity of their relationships.
+   
+   The more complex an architecture is, the harder it will be to extend or to modify; this, in turn, inhibits our ability to extend WASM-GC with new features.
+   
+   We prioriterize simple, composeable architectures over solution-oriented architectures. This will help to ensure that WASM-GC can evolove and be part of future WASM evolution.
+   
+   Our reasoning here is that performance is often the result of progressive tuning of an implementation. It can typically be improved over time. On the other hand, architectural deficiencies are very difficult to correct.
+      
+1. Performance is more important than convenience.
+
+   Since WASM-GC is not intended as a user-facing technology, convenience here refers to the ease with which toolchain authors can target WASM-GC. 
+   
+   Convenience for toolchain authors is still important however, as that affects adoption of WASM-GC. 
+
+   
+## Critical Success Factors
+
+A critical success factor is an aspect or property of possible solutions that may or may not be directly focused on the primary objective, but none-the-less is estimated to be crucially important to the success of the effort.
+
+1. Permit ‘cycle detection’ between host structures and language structures; so that such cycles can be collected if there are no other references to them.
+
+   When a WASM-GC module accesses host capabilities, or when a host application access structures from a WASM-GC library, cross references between them are likely to be established. For example, a DOM node may have an event listener that is actually implemented in WASM. The event listener, in turn, may have a reference to the DOM node it is listening to. This represents a cycle between the host and WASM-GC. If there are no other references to this cycles (if, for example, the DOM node is removed) then the entire cycle must be able to be collected.
+
+1. Performance implications of WASM-GC code should be straightforward.
+
+   A general principle behind WASM itself is that performance should not require special implementation techniques on the part of the engine. I.e., it should not be required that the engine discover particular opportunities for improving performance. In particular, it is the responsibility of languages implementers to generate WASM-GC code that embodies performant strategies.
+   
+   For example, if a type cast is potentially eliminatable, then the engine should not be required to discover that fact; rather the toolchain should be able to eliminate the type cast.
```

Comment:
I generally agree that it is up to language to perform language-level optimizations and thus generate Wasm code that is performant. However I think the way this is phrased (in particular using the word "implementers" versus "implementations" and "toolchain" vs "runtime") gives the impression that this can only be done offline. Whereas I fully expect there will a lot of action building dynamic runtimes on top of Wasm that both dynamically generate code and do dynamic optimizations. So I would propose rephrasing this part a bit to reflect that.

Reply:
Don't disagree (agree even!). However, some of the examples you raise (JITing) are orthogonal to GC. Will attempt to rephrase.